HomeKit: add Thread device

c0ffeeca7 opened the pull request:
## Proposed change
<!-- 
    Describe the big picture of your changes here to communicate to the
    maintainers why we should accept this pull request. If it fixes a bug
    or resolves a feature request, be sure to link to that issue in the 
    additional information section.
-->
HomeKit controller: add to Thread network

- Add procedure and prerequisites
- Remove section on adding a Thread device using a HomePod as a border router. These Apple devices do not communicate their Thread credentials and can therefore currently not be supported.

## Type of change
<!--
    What types of changes does your PR introduce to our documentation/website?
    Put an `x` in the boxes that apply. You can also fill these out after
    creating the PR.
-->

- [ ] Spelling, grammar or other readability improvements (`current` branch).
- [x] Adjusted missing or incorrect information in the current documentation (`current` branch).
- [ ] Added documentation for a new integration I'm adding to Home Assistant (`next` branch).
  - [ ] I've opened up a PR to add logo's and icons in [Brands repository](https://github.com/home-assistant/brands).
- [ ] Added documentation for a new feature I'm adding to Home Assistant (`next` branch).
- [x] Removed stale or deprecated documentation.

## Additional information
<!--
    Details are important, and help maintainers processing your PR.
    Please be sure to fill out additional details, if applicable.
-->

- Link to parent pull request in the codebase: 
- Link to parent pull request in the Brands repository: 
- This PR fixes or closes issue: fixes #

## Checklist
<!--
    Put an `x` in the boxes that apply. You can also fill these out after
    creating the PR. If you're unsure about any of them, don't hesitate to ask.
    We're here to help! This is simply a reminder of what we are going to look
    for before merging your code.
-->

- [x] This PR uses the correct branch, based on one of the following:
  - I made a change to the existing documentation and used the `current` branch.
  - I made a change that is related to an upcoming version of Home Assistant and used the `next` branch.
- [x] The documentation follows the Home Assistant documentation [standards].

[standards]: https://developers.home-assistant.io/docs/documenting/standards

c0ffeeca7 committed:
Draft on adding thread devices.

c0ffeeca7 committed:
HomeKit controller: add to Thread network

- Add prerequisites and procedure.
- Remove section on how to add HomeKit device to Thread network via Apple border router devices such as HomePod mini.
- These devices are not currently supported because they do not communicate their Thread credentials.

c0ffeeca7 committed:
Improve structure

Jc2k commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -83,13 +83,34 @@ HomeKit controller will poll your devices, but it will also automatically enable
+### Prerequisites
+
+* Make sure your HomeKit device supports Thread. This is indicated by the Thread label on the packaging.
+* **Bluetooth support**: HomeKit devices are joined using Bluetooth. If your Home Assistant server device does not support Bluetooth, you can use a Bluetooth proxy.
```
Is it worth calling out the range specifically here? Something like this?

> If your Home Assistant device is too far away from the device you are trying to pair, you can use a bluetooth proxy.

I don't know the brands but I do know some bluetooth/thread locks actually underpower their bluetooth raido as some kind of security feature. We've had reports that people had to put their HA Pi withing 12" for it to work.

agners replied:
There is already a "Bluetooth device support" section above.

Since all HomeKit devices which support Thread do support Bluetooth as well, and a device needs to be paired through Bluetooth first, I'd just reference to that section.

```suggestion
* **Bluetooth support**: The HomeKit device needs to be already connected through Bluetooth (see above).
```

Jc2k commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -83,13 +83,34 @@ HomeKit controller will poll your devices, but it will also automatically enable
+    * If you have a Home Assistant Yellow or SkyConnect, you can enable multiprotocol to set up an Open Thread border router and with that a Thread network. 
+      * Documentation on [enabling multiprotocol on Yellow](https://yellow.home-assistant.io/guides/enable-multiprotocol/)
+      * Documentation on [enabling multiprotocol on SkyConnect](https://skyconnect.home-assistant.io/procedures/enable-multiprotocol/)
+    * Other devices, such as an Apple HomePod mini or Apple TV, cannot currently be used as border routers. They do not communicate their Thread credentials an can therefore not be integrated.
```
They can using the original pairing process so that iOS shares thread credentials for the Apple thread network with the HomeKit devices for us, then they pair to HA.

Jc2k commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -83,13 +83,34 @@ HomeKit controller will poll your devices, but it will also automatically enable
+
+### To add a HomeKit controller device to a Thread network
+
+1. Power up your HomeKit controller device.
```
For this pairing procedure, it is probably a good idea to encourage them to do a full reset if the device has previously been used with thread. Otherwise the device might remember the thread credentials of a different network, and this process might not work.

I think its mentioned elsewhere on this page, but might be worth stressing that if a device isn't discovered, make sure its not still paired to iOS. HomeKit devices can only be paired to a single controller at once. A full reset should avoid this case too.

Jc2k commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -83,13 +83,34 @@ HomeKit controller will poll your devices, but it will also automatically enable
+1. Under **Diagnostic**, you can see the **Thread Status** as **Disabled**.
+    ![Device configuration page](/images/integrations/homekit_controller/homekit_controller_add_02.png)
+1. To enable Thread, under **Configuration**, select **Press**. This will provision the preferred Thread credentials.
+   * The status has now changed to **Child**. 
```
Depending on the device and the mesh size and health, this can also be "Router" and i think also "Leader".

Jc2k commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -83,13 +83,34 @@ HomeKit controller will poll your devices, but it will also automatically enable
-
-HomeKit Controller will automatically discover supported Thread devices using the [Zeroconf](/integrations/zeroconf) integration. Battery powered devices may go to sleep and require a button pressing to wake them up before pairing works.
-
-In order to provision a Thread device onto a mesh network using a HomePod as a border router, first pair it with an iOS device. Then unpair it from the Home app. Don't reset it. This will leave the Thread network details on the device. Home Assistant can now pair with it over Thread.
```
We should try to retain this information, though it should be reworded as it is written as though it is the only configuration procedure which is no longer true. It should probably come after the new procedure as it is a special case.

If you use this original procedure, Apple border routers are absolutely supported. What happens is iOS configures them to join the thread network rather than HA, so HA doesn't need to know the thread credentials for it to work.

One advantage it has over the new procedure is that it is easier to move your iPhone to a device that is screwed to the wall (like an Eve blind) than it is your HA instance. Given bluetooth range limitations (especially for people that haven't invested in a long range bluetooth dongle), this might be the original process is actually better for some users.

agners replied:
This method is now kinda completely missing.

Since currently we can't import HomeKit network credentials to HA, this is the only way to add a HomeKit device through Apple Border Routers to HA.

Under the "Thread device support" section we need two sections:

1. Provision Thread device through Home Assistant (essentially what you've added now)
2. Provision Thread device through Apple Home app (currently documented method)

Jc2k commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -83,13 +83,34 @@ HomeKit controller will poll your devices, but it will also automatically enable
+   * The status has now changed to **Child**. 
+     ![Thread status](/images/integrations/homekit_controller/homekit_controller_add_02.png)
+   * That's it. Your controller device now communicates via Thread.
+   * Your HomeKit controller automatically discovers supported Thread devices using the [Zeroconf](/integrations/zeroconf) integration. Battery powered devices may go to sleep. To pair them, you may need to press a button to wake them up.
```
In my own testing, this really only seems to apply to devices that are on a thread network but unpaired. Pairing over bluetooth first seems to avoid it, so we can probably drop that comment here.

agners replied:
This is related to the other method to on-board a Thread device, so it doesn't belong here.

```suggestion
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -83,13 +83,34 @@ HomeKit controller will poll your devices, but it will also automatically enable
-HomeKit Controller will automatically discover supported Thread devices using the [Zeroconf](/integrations/zeroconf) integration. Battery powered devices may go to sleep and require a button pressing to wake them up before pairing works.
-
-In order to provision a Thread device onto a mesh network using a HomePod as a border router, first pair it with an iOS device. Then unpair it from the Home app. Don't reset it. This will leave the Thread network details on the device. Home Assistant can now pair with it over Thread.
+## Adding a HomeKit controller device to a Thread network
```
This is somewhat confusing. Since this is already in the HomeKit Controller integration, I'd go for

```suggestion
## Adding a HomeKit device to a Thread network
```

Jc2k commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -83,13 +83,34 @@ HomeKit controller will poll your devices, but it will also automatically enable
+* Make sure your HomeKit device supports Thread. This is indicated by the Thread label on the packaging.
+* **Bluetooth support**: HomeKit devices are joined using Bluetooth. If your Home Assistant server device does not support Bluetooth, you can use a Bluetooth proxy.
+* **Thread network**: In order to use HomeKit over Thread you need a working border router. 
+    * Make sure your Home Assistant device is on the same network (LAN) as the border router. 
```
Should we call out common mistakes here? Or reference a shared thread page for that? (E.g. Cross VLAN not supported, IPv6 must be set to auto in HAOS, i can think of more if this is the right place for that information)

c0ffeeca7 replied:
Let's get the procedures done first. Then we can add troubleshooting items in a separate PR.

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -83,13 +83,34 @@ HomeKit controller will poll your devices, but it will also automatically enable
+   * If you have Bluetooth enabled, it should be discovered under **{% my integrations title="Settings > Devices & Services" %}**.
+1. On the HomeKit integration, select **Configure**.
+1. To pair the device, enter the HomeKit pairing code. The code is on the device itself or the packaging. 
+   * Select **Finish**. 
```
Since we make Bluetooth connection a prerequisite, this won't be necessary.

```suggestion
```

We could move this steps to the Bluetooth section though :thinking:

Jc2k reviewed (dismissed):
Thanks! This is going to be so much more useful.

I've noted a few things that come up in support a bit, but I don't know if its best to reference them here or in shared thread documentation.

The biggie for me is I want to retain (and improve?) the existing documentation on how to use HomeKit Thread devices with Apple border routers.

Currently that is this sad little paragraph i wrote:

> In order to provision a Thread device onto a mesh network using a HomePod as a border router, first pair it with an iOS device. Then unpair it from the Home app. Don't reset it. This will leave the Thread network details on the device. Home Assistant can now pair with it over Thread.

We have lots of users using this approach, and since HAOS 10 there are some with e.g. 10 homepod's with multiple Eve devices connected to it.

c0ffeeca7 committed:
Review feedback: move Bluetooth steps into separate section

c0ffeeca7 committed:
Review comment Jc2K: remove section

- comment: "In my own testing, this really only seems to apply to devices that are on a thread network but unpaired. ..."

c0ffeeca7 committed:
Review comment: add other Thread statuses

c0ffeeca7 committed:
Review comment: add info to reset device before

c0ffeeca7 committed:
Add image of HomeKit integration

c0ffeeca7 committed:
Rev comment: Apple devices cannot be used as OTBR: remove info

c0ffeeca7 committed:
Add HomeKit device to Thread network via Apple OTBR

- Add procedure
- Cleanup terminology in the Thread section

c0ffeeca7 committed:
Add introductions

c0ffeeca7 commented:
Thank you for your thorough review! I believe I've implemented almost all of it. 
The section about troubleshooting the LAN: I'd prefer to do this in a separate PR.

c0ffeeca7 committed:
Fix markdown

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,82 @@ HomeKit IP accessories for these device types may work with some caveats:
+### Prerequisites
 
-In order to use HomeKit over Thread you need a working border router (like a HomePod mini). Your Home Assistant instance will need to be on the same VLAN as the border router.
+- If your Home Assistant instance does not natively support Bluetooth, use a Bluetooth proxy. 
```
```suggestion
- If your Home Assistant instance does not natively support Bluetooth, use an ESPHome Bluetooth proxy. 
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,82 @@ HomeKit IP accessories for these device types may work with some caveats:
+  - A proxy can also be helpful if your Home Assistant device is too far away from the device you are trying to pair.
+- If your HomeKit device has been used with Thread before, or is still paired with iOS, reset the device.
+   - HomeKit devices can only be paired to a single controller at once. 
+   - If it has been in a Thread network before, the device might remember the Thread credentials of a different network. A reset fixes that.
```
```suggestion
   - If it has been in a Thread network before, the device might remember the Thread credentials of a different network. A reset makes sure the device is not connected to any Thread network.
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,82 @@ HomeKit IP accessories for these device types may work with some caveats:
+   - Select **Finish**. 
+1. The HomeKit device now automatically discovers other Bluetooth devices (if the [Bluetooth](/integrations/bluetooth) integration is enabled).  
+   - Bluetooth devices may take significantly longer to pair than IP devices. 
+   - To pair a battery-powered device, you may need to press a button on the device to wake it.
```
This doesn't belong here anymore, as we have the Bluetooth integration as a prerequisite.

```suggestion
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,82 @@ HomeKit IP accessories for these device types may work with some caveats:
+   - If you have Bluetooth enabled, the device should be discovered under **{% my integrations title="Settings > Devices & Services" %}**.
+1. On the HomeKit integration, select **Configure**.
+     ![HomeKit integration](/images/integrations/homekit_controller/homekit_controller_add_01.png)
+1. To pair the device, enter the HomeKit pairing code. The code is on the device itself or the packaging. 
```
Information from the step below:

```suggestion
1. To pair the device, enter the HomeKit pairing code. The code is on the device itself or the packaging. 
   - To pair a battery-powered device, you may need to press a button on the device to wake it.
   - Bluetooth devices may take significantly longer to pair than IP devices.
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,82 @@ HomeKit IP accessories for these device types may work with some caveats:
+## Adding a HomeKit device through Bluetooth
 
-HomeKit Controller will automatically discover Bluetooth devices once the [Bluetooth](/integrations/bluetooth) integration is enabled and functional. Bluetooth devices may take significantly longer to pair than IP devices. Battery-powered devices may require pressing a button on the device to wake it before pairing can be successful.
+You can add a HomeKit device to Home Assistant via Bluetooth.
```
```suggestion
You can add a HomeKit device to Home Assistant via [Bluetooth](/integrations/bluetooth).
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,82 @@ HomeKit IP accessories for these device types may work with some caveats:
+### Prerequisites
+
+- A HomeKit device which supports Thread. This is indicated by the Thread label on the packaging.
+- Make sure HomeKit controller has been [joined using Bluetooth](#adding-a-homekit-device-through-bluetooth). 
```
```suggestion
- Make sure the HomeKit device has been [joined using Bluetooth](#adding-a-homekit-device-through-bluetooth). 
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,82 @@ HomeKit IP accessories for these device types may work with some caveats:
+- A HomeKit device which supports Thread. This is indicated by the Thread label on the packaging.
+- Make sure HomeKit controller has been [joined using Bluetooth](#adding-a-homekit-device-through-bluetooth). 
+- **Thread network**: In order to use HomeKit over Thread, you need a working border router. 
+    - Make sure your Home Assistant device is on the same network (LAN) as the border router.
```
```suggestion
    - Make sure your Home Assistant device is on the same network (LAN) as the border router.
    - Make sure the Thread network you'd like to use is known by Home Assistant and marked as **Preferred network** in the Thread configuration.
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,82 @@ HomeKit IP accessories for these device types may work with some caveats:
+      - Documentation on [enabling multiprotocol on Yellow](https://yellow.home-assistant.io/guides/enable-multiprotocol/)
+      - Documentation on [enabling multiprotocol on SkyConnect](https://skyconnect.home-assistant.io/procedures/enable-multiprotocol/)
+
+### To add a HomeKit  device to a Thread network via Home Assistant
```
```suggestion
### To add a HomeKit device to a Thread network via Home Assistant
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,82 @@ HomeKit IP accessories for these device types may work with some caveats:
+1. Unpair the HomeKit device from the Apple Home app. Don't reset the device.
+   - This leaves the Thread network details on the HomeKit device.
+1. The device will be automatically discovered by the HomeKit controller integration in Home Assistant. 
+   - It will appear as a discovered device over Thread.
```
Just like Bluetooth, you'll have to click configure here. Then it will be added.

c0ffeeca7 replied:
Thank you for the catch! Added missing steps.

c0ffeeca7 committed:
Apply suggestions from code review

Co-authored-by: Stefan Agner <stefan@agner.ch>

c0ffeeca7 committed:
Add to Thread via Apple border router

- add missing steps to procedure

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,95 @@ HomeKit IP accessories for these device types may work with some caveats:
+
+There are two methods to add a HomeKit device to a Thread network:
+
+- via Home Assistant
```
"via Home Assistant" on its own is a bit unclear to me. I'd describe it a bit more. The title can still be "via Home Assistant" only.

```suggestion
- via Home Assistant's preferred Thread network
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,95 @@ HomeKit IP accessories for these device types may work with some caveats:
+
+There are two methods to add a HomeKit device to a Thread network:
+
+- via [Home Assistant](#adding-a-homekit-device-to-a-thread-network-via-home-assistant)
```
As above

```suggestion
- via [Home Assistant's preferred Thread network](#adding-a-homekit-device-to-a-thread-network-via-home-assistant)
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,95 @@ HomeKit IP accessories for these device types may work with some caveats:
+- via Home Assistant
+- via [Apple Thread border router](#adding-a-homekit-device-to-home-assistant-via-apple-thread-border-router)
+
+This section describes how to add it via Home Assistant.
```
```suggestion
This section describes how to add it via Home Assistant's preferred Thread network.
```

agners commented on source/_integrations/homekit_controller.markdown:
```diff
@@ -79,17 +79,95 @@ HomeKit IP accessories for these device types may work with some caveats:
+- An Apple device that can act as a Thread border router, such as a HomePod mini.
+- A HomeKit device which supports Thread. This is indicated by the Thread label on the packaging.
+- Make sure your Home Assistant instance is on the same network (LAN) as the border router.
+- Make sure the HomeKit device has been paired in the Apple Home app (using an iOS app).
```
```suggestion
- Make sure the HomeKit device has been paired in the Apple Home app (using the iOS Home app).
```

c0ffeeca7 committed:
Apply suggestions from code review

Co-authored-by: Stefan Agner <stefan@agner.ch>

Jc2k commented:
We were still discussing this on discord. Sorry I should have added a dummy review comment to warn you.

frenck commented:
I'm sorry, didn't know.

Options are either to revert this, or, iterate on it.

../Frenck

Jc2k commented:
No my bad I though it wouldn't be merged until I'd acked. Should have been more explicit.

This creates a new top level thread and Bluetooth section and sort of orphans the WiFi/Ethernet stuff a bit. And they are probably still the most important/used device types. So the top of the page needs reworking a bit now that we've got these detailed and awesome notes for thread/Bluetooth so that it flows properly (my writing style isn't as good!), and the "ip devices" notes need to become a "pairing a wifi or Ethernet homekit device" section above the Bluetooth one or something.

I was hoping it was ok enough as is (and iterate on current, consolidate the WiFi/Ethernet in follow up pr). But I wanted to make sure it was still clear that WiFi/Ethernet is actually supported while that iteration happens.

I wanted to check how it read not in diff view on my phone, but in a dev env, it's a bit sizeable to review as diff hunks on my phone and I haven't been near my Ha dev box this week 🥲

If you think it's clear that "original" homekit wifi and Ethernet devices are still supported I'm happy to iterate on current.

c0ffeeca7 commented:
The error is mine. I should have marked this as draft.
